c10/util/SymIntTable.cpp

```diff
@@ -0,0 +1,34 @@
+#include <ATen/core/SymbolicIntNode.h>
+
+
+namespace c10 {
+
+    size_t SymIntTable::addNode(SymbolicIntNode* sin) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        auto index = nodes_.size();
+        nodes_.push_back(sin);
+        return index;
+    }
+    SymbolicIntNode* SymIntTable::getNode(size_t index) {
+        TORCH_CHECK(index < nodes_.size());
+        std::lock_guard<std::mutex> lock(mutex_);
+        return nodes_[index];
+    }
+
+    c10::SymInt SymbolicIntNode::toSymInt() {
+        //TODO: memoize this
+        auto& sit = getSymIntTable();
+        auto data = sit.addNode(this) | SYM_TAG_MASK;
+        return c10::SymInt(data);
```

Comment:
I wouldn't call the transformation you need to do here memoization; instead, the invariant is that SymbolicIntNodes should only ever be allocated SIMULTANEOUSLY with assigning them in the table, so they know their index in that case.